interop-cats/jvm/src/main/scala/scalaz/zio/interop/stream/Stream.scala

```diff
@@ -0,0 +1,157 @@
+ * limitations under the License.
+ */
+
+package scalaz.zio.interop.stream
+
+import cats.~>
+import cats.effect.Effect
+import cats.implicits._
+import scalaz.zio.{Chunk, Queue, ZIO, ZManaged, ZSchedule}
+import scalaz.zio.stream.{Take, ZSink, ZStream}
+
+final class Stream[F[+ _], +A] private[stream] (private[stream] val underlying: ZStream[Any, Throwable, A])(
+  implicit liftIO: ZIO[Any, Throwable, ?] ~> F,
+  liftZIO: F ~> ZIO[Any, Throwable, ?],
+  effect: Effect[F]
+) { self =>
+
+  import Stream.Fold
+
+  def fold[A1 >: A, S]: Fold[F, A1, S] = ???
+
+  def foldLeft[A1 >: A, S](s: S)(f: (S, A1) => F[S]): F[S] = {
+    (fold[A, S]: F[(S, S => Boolean, (S, A) => F[S]) => F[S]]).flatMap(_(s, _ => false, f))
+  }
+
+  def ++[A1 >: A](that: => Stream[F, A1]): Stream[F, A1] =
+    new Stream(underlying ++ that.underlying)
+
+  def drain: Stream[F, Nothing] =
+    new Stream(underlying.drain)
+
+  def filter(pred: A => Boolean): Stream[F, A] =
+    new Stream(underlying.filter(pred))
+
+  def filterM(pred: A => F[Boolean]): Stream[F, A] =
+    new Stream(underlying.filterM(pred.andThen(liftZIO.apply)))
+
+  def filterNot(pred: A => Boolean): Stream[F, A] =
+    filter(a => !pred(a))
+
+  def foreachWhile(f: A => F[Boolean]): F[Unit] =
+    liftIO.apply(underlying.foreachWhile(f.andThen(liftZIO.apply)))
+
+  def collect[B](pf: PartialFunction[A, B]): Stream[F, B] =
+    new Stream(underlying.collect(pf))
+
+  def drop(n: Int): Stream[F, A] =
+    new Stream(underlying.drop(n))
+
+  def dropWhile(pred: A => Boolean): Stream[F, A] =
+    new Stream(underlying.dropWhile(pred))
+
+  def flatMap[B](f: A => Stream[F, B]): Stream[F, B] =
+    new Stream(underlying.flatMap(f.andThen(_.underlying)))
+
+  def foreach(f: A => F[Unit]): F[Unit] =
+    liftIO(underlying.foreach(f.andThen(liftZIO.apply)))
+
+  def forever: Stream[F, A] =
+    new Stream(underlying.forever)
+
+  def map[B](f: A => B): Stream[F, B] =
+    new Stream(underlying.map(f))
+
+  // TODO Should we expose Chunk to clients?
+  def mapConcat[B](f: A => Chunk[B]): Stream[F, B] =
+    new Stream(underlying.mapConcat(f))
+
+  def mapM[B](f: A => F[B]): Stream[F, B] =
+    new Stream(underlying.mapM(f.andThen(liftZIO.apply)))
+
+  def merge[A1 >: A](that: Stream[F, A1], capacity: Int = 1): Stream[F, A1] =
+    new Stream(underlying.merge(that.underlying, capacity))
+
+  def mergeEither[B](that: Stream[F, B], capacity: Int = 1): Stream[F, Either[A, B]] =
+    new Stream(underlying.mergeEither(that.underlying, capacity))
+
+  def mergeWith[B, C](that: Stream[F, B], capacity: Int = 1)(l: A => C, r: B => C): Stream[F, C] =
+    new Stream(underlying.mergeWith(that.underlying, capacity)(l, r))
+
+  // TODO Sink, Managed
+  def peel[A1 >: A, B](sink: ZSink[Any, Throwable, A1, A1, B]): ZManaged[Any, Throwable, (B, Stream[F, A1])] = ???
+
+  // TODO Schedule #790
+  //  Find out how to deal correctly with "with Clock"
+  def repeat(schedule: ZSchedule[Any, Unit, _]): Stream[F, A] = ???
```

Comment:
With the new representation suggested above, you can just require `Runtime[R]`.

Actually, it seems to me that we don't need `Runtime[R]` for every method. We can get away with requiring it on:

 - `def run`
 - `def foreach`
 - Any other methods that "escape" `Stream`.

This would push the `Clock` constraint (and others) to `underlying: ZStream[Clock with ..., E, A]`.

Reply:
Let me know if this is unclear and I'll explain more. 😄

Reply:
I've managed to do this by adding a method `provide` to `ZStream`, not sure this is the way to go but it seems quite elegant.